### PR TITLE
Implementation for BNB Issue#1

### PR DESCRIPTION
**Pull request by NeoZ666**

### CODE REVIEW

**Comment by delcin-raj** [comment]
make sure to run "cargo fmt" before committing changes

**Comment by NeoZ666** [comment]
Done

**Comment by NeoZ666** [comment]
Ran the `cargo fmt` command once again, please check.

File: src/lib.rs [hunk 1]
```diff
@@ -96,20 +97,146 @@ pub struct SelectionOutput {
 pub fn select_coin_bnb(
     inputs: &[OutputGroup],
     options: CoinSelectionOpt,
+    rng: &mut ThreadRng,
 ) -> Result<SelectionOutput, SelectionError> {
-    unimplemented!()
+    let mut selected_inputs: Vec<usize> = vec![];
+    let bnb_tries = 1000000;
```
Comment (delcin-raj) [inline src/lib.rs:103]: this should be a constant.

File: src/lib.rs [hunk 2]
```diff
@@ -96,20 +97,146 @@ pub struct SelectionOutput {
...
+                accumulated_weight,
+                estimated_fee,
+            );
+            let selection_output = SelectionOutput {
+                selected_inputs: selected_coin,
+                waste: WasteMetric(waste),
+            };
+            Ok(selection_output)
+        }
+        None => select_coin_srd(inputs, options, &mut rand::thread_rng()),
```
Comment (delcin-raj) [inline src/lib.rs:144]: Should be Err(NoSolutionFound)

File: src/lib.rs [hunk 3]
```diff
@@ -96,20 +97,146 @@ pub struct SelectionOutput {
...
+) -> Option<Vec<usize>> {
+    let target_for_match = options.target_value
+        + calculate_fee(options.base_weight, options.target_feerate)
+        + options.cost_per_output;
+    let match_range = options.cost_per_input + options.cost_per_output;
+    if acc_eff_value > target_for_match + match_range {
+        return None;
+    } else if acc_eff_value >= target_for_match {
+        return Some(selected_inputs.to_vec());
+    } else if bnp_tries <= 0 || depth >= inputs_in_desc_value.len() {
```
Comment (delcin-raj) [inline src/lib.rs:167]: should be `bnp_tries == 0` or `bnp_tries < 1` because it is unsigned and can never be less than 0.

File: src/lib.rs (hunk 2 above)
Comment (delcin-raj) [reply src/lib.rs:144]: The reasoning behind this is when the user calls the high level API, srd will be invoked regardless of the result of bnb.
So when a user calls bnb in particular signalling no solution found is better than returning the result of srd

File: src/lib.rs [hunk 4]
```diff
@@ -96,20 +97,146 @@ pub struct SelectionOutput {
...
+            match with_this {
+                Some(_) => return with_this,
+                None => {
+                    selected_inputs.pop(); //poping out the selected utxo if it does not fit
+                    let without_this = bnb(
+                        inputs_in_desc_value,
+                        selected_inputs,
+                        acc_eff_value,
+                        depth + 1,
+                        bnp_tries - 1,
```
Comment (delcin-raj) [inline src/lib.rs:194]: bnp_tries - 2

File: src/lib.rs [hunk 5]
```diff
@@ -96,20 +97,146 @@ pub struct SelectionOutput {
...
+                None => {
+                    let new_effective_values = acc_eff_value
+                        + effective_value(&inputs_in_desc_value[depth].1, options.target_feerate);
+                    selected_inputs.push(inputs_in_desc_value[depth].0);
+                    let with_this = bnb(
+                        inputs_in_desc_value,
+                        selected_inputs,
+                        new_effective_values,
+                        depth + 1,
+                        bnp_tries - 1,
```
Comment (delcin-raj) [inline src/lib.rs:225]: bnp_tries - 2

File: src/lib.rs (hunk 4 above)
Comment (delcin-raj) [reply src/lib.rs:194]: because every recursive call should decrement this by one.
This variable is immutable

File: src/lib.rs [hunk 6]
```diff
@@ -358,6 +483,11 @@ fn effective_value(output: &OutputGroup, feerate: f32) -> u64 {
         .saturating_sub(calculate_fee(output.weight, feerate))
 }
 
+fn generate_random_bool(rng: &mut ThreadRng) -> bool {
```
Comment (delcin-raj) [inline src/lib.rs:486]: Remove this function

File: src/lib.rs [hunk 7]
```diff
@@ -248,17 +375,15 @@ pub fn select_coin_fifo(
 pub fn select_coin_srd(
```
Comment (delcin-raj) [inline src/lib.rs:375]: Undo the changes to this function

File: src/lib.rs [hunk 8]
```diff
@@ -541,7 +706,7 @@ mod test {
     fn test_insufficient_funds() {
         let inputs = setup_basic_output_groups();
         let options = setup_options(7000); // Set a target value higher than the sum of all inputs
-        let result = select_coin_srd(&inputs, options);
+        let result = select_coin_srd(&inputs, options, &mut rand::thread_rng());
```
Comment (delcin-raj) [inline src/lib.rs:709]: undo the changes

File: src/lib.rs [hunk 9]
```diff
@@ -520,12 +650,47 @@ mod test {
...
+        assert!(ans.is_ok());
+        assert!(!ans.unwrap().selected_inputs.contains(&0));
+        // as 10000000 should not be included in the selection
     }
 
     fn test_successful_selection() {
         let mut inputs = setup_basic_output_groups();
         let mut options = setup_options(2500);
-        let mut result = select_coin_srd(&inputs, options);
+        let mut result = select_coin_srd(&inputs, options, &mut rand::thread_rng());
```
Comment (delcin-raj) [inline src/lib.rs:693]: undo the changes

**Review by delcin-raj — COMMENTED** [review]
Add tests from core

File: src/lib.rs [hunk 10]
```diff
@@ -96,20 +97,146 @@ pub struct SelectionOutput {
...
+        + calculate_fee(options.base_weight, options.target_feerate)
+        + options.cost_per_output;
+    let match_range = options.cost_per_input + options.cost_per_output;
+    if acc_eff_value > target_for_match + match_range {
+        return None;
+    } else if acc_eff_value >= target_for_match {
+        return Some(selected_inputs.to_vec());
+    } else if bnp_tries <= 0 || depth >= inputs_in_desc_value.len() {
+        return None;
+    } else {
```
Comment (aruokhai) [inline src/lib.rs:169]: A series of `if clauses` without the usage of` if-else ` clause would be more readable and also the `else` clause is redundant.

File: src/lib.rs [hunk 11]
```diff
@@ -102,20 +104,148 @@ pub struct SelectionOutput {
...
+    selected_inputs: &mut Vec<usize>,
+    acc_eff_value: u64,
     depth: usize,
     bnp_tries: u32,
     options: &CoinSelectionOpt,
-) -> Vec<usize> {
-    unimplemented!()
+    rng: &mut ThreadRng,
+) -> Option<Vec<usize>> {
+    let target_for_match = options.target_value
```
Comment (delcin-raj) [inline src/lib.rs:167]: This value could be computed only once as it depends on the options
No need to recompute this during every recursive call.
This can be included in the function parameter

File: src/lib.rs [hunk 12]
```diff
@@ -102,20 +104,148 @@ pub struct SelectionOutput {
...
+        + calculate_fee(options.base_weight, options.target_feerate)
+        + options.cost_per_output;
+    let match_range = options.cost_per_input + options.cost_per_output;
```
Comment (delcin-raj) [inline src/lib.rs:170]: This also

File: src/lib.rs [hunk 13]
```diff
@@ -102,20 +104,148 @@ pub struct SelectionOutput {
...
+    if acc_eff_value >= target_for_match {
+        return Some(selected_inputs.to_vec());
+    }
+    if bnp_tries == 0 || depth >= inputs_in_desc_value.len() {
+        return None;
+    }
+    if rng.gen_bool(0.5) {
+        // exploring the inclusion branch
+        // first include then omit
+        let new_effective_values =
```
Comment (delcin-raj) [inline src/lib.rs:183]: it's a single value

File: src/lib.rs [hunk 14]
```diff
@@ -102,20 +104,148 @@ pub struct SelectionOutput {
...
+// changing the selected_inputs : &[usize] -> &mut Vec<usize>
 fn bnb(
     inputs_in_desc_value: &[(usize, OutputGroup)],
-    selected_inputs: &[usize],
-    effective_value: u64,
+    selected_inputs: &mut Vec<usize>,
+    acc_eff_value: u64,
     depth: usize,
     bnp_tries: u32,
     options: &CoinSelectionOpt,
```
Comment (delcin-raj) [inline src/lib.rs:164]: This argument `options` could potentially be discarded

File: src/lib.rs [hunk 15]
```diff
@@ -102,20 +104,148 @@ pub struct SelectionOutput {
...
+            depth + 1,
+            bnp_tries - 1,
+            options,
+            rng,
+        );
+        match with_this {
+            Some(_) => with_this,
+            None => {
+                selected_inputs.pop(); //poping out the selected utxo if it does not fit
+                let without_this = bnb(
```
Comment (delcin-raj) [inline src/lib.rs:199]: Just return the result of the `bnb` call here.
No need for `without_this` variable

File: src/lib.rs [hunk 16]
```diff
@@ -102,20 +104,148 @@ pub struct SelectionOutput {
...
 /// Return empty vec if no solutions are found
+// changing the selected_inputs : &[usize] -> &mut Vec<usize>
 fn bnb(
     inputs_in_desc_value: &[(usize, OutputGroup)],
-    selected_inputs: &[usize],
-    effective_value: u64,
+    selected_inputs: &mut Vec<usize>,
+    acc_eff_value: u64,
     depth: usize,
     bnp_tries: u32,
```
Comment (delcin-raj) [inline src/lib.rs:163]: Correct it to `bnb_tries`

**Review by delcin-raj — CHANGES_REQUESTED** [review]

**Review by delcin-raj — CHANGES_REQUESTED** [review]

**Comment by delcin-raj** [comment]
Waiting for the updates...

File: src/lib.rs [hunk 17]
```diff
@@ -98,25 +98,177 @@ pub struct SelectionOutput {
...
+                    acc_eff_value,
+                    depth + 1,
+                    bnb_tries,
+                    rng,
+                    match_parameters,
+                )
+            }
+        }
+    } else {
+        /// Avoided creation of intermediate variable, for marginally better performance.
```
Comment (delcin-raj) [inline src/lib.rs:235]: remove this comment, confusing to the people who don't have context

File: src/lib.rs [hunk 18]
```diff
@@ -98,25 +98,177 @@ pub struct SelectionOutput {
...
+                    depth + 1,
+                    bnb_tries,
+                    rng,
+                    match_parameters,
+                );
+                match with_this {
+                    Some(_) => with_this,
+                    None => {
+                        selected_inputs.pop(); // poping out the selected utxo if it does not fit
+                        None // this may or may not be correct
```
Comment (delcin-raj) [inline src/lib.rs:266]: This is correct, so remove the comment

**Review by delcin-raj — CHANGES_REQUESTED** [review]

File: src/lib.rs (hunk 17 above)
Comment (NeoZ666) [reply src/lib.rs:235]: done

File: src/lib.rs [hunk 19]
```diff
@@ -73,49 +76,191 @@ pub enum ExcessStrategy {
...
+                selected_inputs: selected_coin,
+                waste: WasteMetric(waste),
+            };
+            Ok(selection_output)
+        }
+        None => Err(SelectionError::NoSolutionFound),
+    }
 }
 
 /// Return empty vec if no solutions are found
```
[the remark this answers is not in the document]
Comment (NeoZ666) [reply src/lib.rs:176]: done

**Review by delcin-raj — APPROVED** [review]
LGTM
